File: packages/receipts/src/sdk/openai-agents.ts
// OpenAI Agents SDK (JS) adapter.
//   wrapTools:     enforcement + receipts, by wrapping each FunctionTool's invoke(). A denied call never runs; the
//                  model receives the denial text as the tool result and the run continues.
//   observeRunner: receipts only, from the runner's agent_tool_start / agent_tool_end events. Cannot block, so it
//                  evaluates no policy. Use one or the other per tool, not both.
// Typed structurally so this file does not import the package.
import { receiptIdOf, type SdkIssuer, type ToolEvent } from "./index.ts";

interface ToolCallDetails {
  toolCall?: { callId?: string; arguments?: string };
}
// Parameters are `any` on purpose: the SDK's invoke() takes its own RunContext and ToolCallDetails types, and a
// narrower structural signature here would not be assignable under strictFunctionTypes.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type InvokeFn = (runContext: any, input: string, details?: any) => Promise<unknown>;

function parseArgs(input: string | undefined): Record<string, unknown> {
  if (!input) return {};
  try {
    const v = JSON.parse(input) as unknown;
    return v && typeof v === "object" && !Array.isArray(v) ? (v as Record<string, unknown>) : { input: v };
  } catch {
    return { input };
  }
}

function parseResult(result: unknown): unknown {
  if (typeof result !== "string") return result;
  try {
    return JSON.parse(result) as unknown;
  } catch {
    return result;
  }
}

export function wrapTools<T extends { name: string; invoke: InvokeFn }>(issuer: SdkIssuer, tools: readonly T[]): T[] {
  return tools.map((t) => {
    const invoke: InvokeFn = async (ctx, input, details?: ToolCallDetails) => {
      const ev: ToolEvent = { tool: t.name, args: parseArgs(input), session: { id: null, toolUseId: details?.toolCall?.callId ?? null } };
      const policy = issuer.decide(ev);
      if (policy && policy.decision === "deny") {
        const reason = [...policy.reasons, ...policy.errors].join("; ") || "no permit policy matched";
        const bundle = issuer.record(ev, { status: "denied", reason }, policy);
        return `Denied by policy: ${reason} (receipt ${receiptIdOf(bundle)})`;
      }
      try {
        const result = await t.invoke(ctx, input, details);
        issuer.record(ev, { status: "executed", result: parseResult(result) }, policy);
        return result;
      } catch (e) {
        issuer.record(ev, { status: "error", error: e instanceof Error ? e.message : String(e) }, policy);
        throw e;
      }
    };
    return { ...t, invoke } as T;
  });
}

interface Listenable {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  on(event: any, listener: (...args: any[]) => void): unknown;
}

export function observeRunner(issuer: SdkIssuer, runner: Listenable): void {
  const pending = new Map<string, ToolEvent>();
  runner.on("agent_tool_start", (_ctx: unknown, _agent: unknown, tool: { name: string }, details?: ToolCallDetails) => {
    const callId = details?.toolCall?.callId;
    if (!callId) return;
    pending.set(callId, { tool: tool.name, args: parseArgs(details?.toolCall?.arguments), session: { id: null, toolUseId: callId } });
  });
  runner.on("agent_tool_end", (_ctx: unknown, _agent: unknown, tool: { name: string }, result: unknown, details?: ToolCallDetails) => {
    const callId = details?.toolCall?.callId ?? "";
    const ev = pending.get(callId) ?? { tool: tool.name, args: {}, session: { id: null, toolUseId: callId || null } };
    pending.delete(callId);
    issuer.record(ev, { status: "executed", result: parseResult(result) }, null);
  });
}

File: packages/python/agent_custody/openai_agents.py
"""OpenAI Agents SDK (Python) adapter: wraps each FunctionTool's on_invoke_tool. Decides, runs, records.

A denied call never runs; the model receives the denial text as the tool result and the run continues.
"""
from __future__ import annotations

import dataclasses
import json
from typing import Any, Dict, List

from agents import FunctionTool

from . import Client, receipt_id_of


def _parse(input_json: str) -> Dict[str, Any]:
    try:
        v = json.loads(input_json) if input_json else {}
        return v if isinstance(v, dict) else {"input": v}
    except ValueError:
        return {"input": input_json}


def _result(v: Any) -> Any:
    if not isinstance(v, str):
        return v
    try:
        return json.loads(v)
    except ValueError:
        return v


def wrap_tools(client: Client, tools: List[FunctionTool]) -> List[FunctionTool]:
    out: List[FunctionTool] = []
    for t in tools:
        original = t.on_invoke_tool

        async def invoke(ctx: Any, input_json: str, _t: FunctionTool = t, _orig=original) -> Any:
            args = _parse(input_json)
            policy = client.decide(_t.name, args)
            if policy and policy["decision"] == "deny":
                reason = "; ".join(policy["reasons"] + policy["errors"]) or "no permit policy matched"
                bundle = client.record(_t.name, args, {"status": "denied", "reason": reason}, policy)
                return f"Denied by policy: {reason} (receipt {receipt_id_of(bundle)})"
            try:
                result = await _orig(ctx, input_json)
            except Exception as e:
                client.record(_t.name, args, {"status": "error", "error": str(e)}, policy)
                raise
            client.record(_t.name, args, {"status": "executed", "result": _result(result)}, policy)
            return result

        out.append(dataclasses.replace(t, on_invoke_tool=invoke))
    return out
